### PR TITLE
irmin-pack; add an option to cancel existing freeze instead of blocking

### LINKED ISSUE #1172: Add a "cancel existing freeze" option for irmin-pack layer freezes
A new idea for the freeze strategy. Currently we have tried:

1. (default) the 1st concurrent freeze is blocked, the 2snd freeze blocks all the next writes
2. (memory overcommit) all concurrent freeze are discarded ; but the (long-running) freeze in progress is still competing with concurrent writes and so can take a long time to finish (and so the memory usage is unbounded). See #1156 

Here is a 3rd option:

3. (cancel) there is at most one freeze ; every concurrent freezes cancel the one already running.

The worst case for 3. is that we keep cancelling the freeze (and so there is no GC going on disk) but at least we 1/ don’t block writes (and so keep a good availability) and 2/ we don’t keep an unbounded “newies” list in memory (and so we don’t have unbouded memory usage).

### PR DESCRIPTION
**Pull request by samoht**
Fix #1172
On top of #1190 

The result:

```
» time dune exec -- ./bench/irmin-pack/layers.exe -n 1000 -v --freeze-throttle=cancel-existing
Running benchmarks in bench/irmin-pack/layers.ml:

  { ncommits = 1000;
    ncycles = 10;
    depth = 10;
    root = "test-bench";
    clear = false;
    no_freeze = false;
    show_stats = false;
    freeze_throttle = Cancel_existing }

+1607624817us [INFO] application exec: rm -rf test-bench

+1607624817us [INFO] index [upper1] v fresh=true readonly=false log_size=2000000
+1607624817us [INFO] index [upper0] v fresh=true readonly=false log_size=2000000
+1607624821us [INFO] irmin-pack freeze started { id=1; max=d6e260afc48483aaa50547b09ac2071704931568; min_upper=27e86e2486477fd8f2e06765224351a7e1da9f80; }
+1607624822us [INFO] irmin-pack cancelled freeze { id=1; waiting=34.452us; duration=1.272s }
+1607624822us [INFO] irmin-pack freeze started { id=2; max=6f26f9d7d46ebc525c06b52ca621aee01515056d; min_upper=02d55ce623efa0c3e6d664c6a601bfc76ff2c003; }
+1607624824us [INFO] irmin-pack cancelled freeze { id=2; waiting=99.325us; duration=1.516s }
+1607624824us [INFO] irmin-pack freeze started { id=3; max=de09f7fc557476f2695a66e07cb8bd05043290f2; min_upper=180ceec0e30833f65b4824f2fbd854b7a4daad29; }
+1607624825us [INFO] irmin-pack cancelled freeze { id=3; waiting=115us; duration=1.267s }
+1607624825us [INFO] irmin-pack freeze started { id=4; max=9f4c1195aa19aaf02b1ca9fb5c35d351b72d10fa; min_upper=b272ea961460d40226f49729229a985520fe5ab4; }
+1607624826us [INFO] irmin-pack cancelled freeze { id=4; waiting=72.132us; duration=1.495s }
+1607624826us [INFO] irmin-pack freeze started { id=5; max=fd3ae9fd86d5aa8c5446c3d44571467d21fcc0fb; min_upper=0d8f0ca3ac4da5566e9379ab28f42b261e82b3bb; }
+1607624828us [INFO] irmin-pack cancelled freeze { id=5; waiting=68.585us; duration=1.455s }
+1607624828us [INFO] irmin-pack freeze started { id=6; max=6c6ae46382f6770a94c9249ddf76caacb540c97e; min_upper=e25eedbdc7a017b5eeba19007fb601a345471ec8; }
+1607624829us [INFO] irmin-pack cancelled freeze { id=6; waiting=71.416us; duration=1.532s }
+1607624829us [INFO] irmin-pack freeze started { id=7; max=faa102ad1ca3de0d5972060dab5ada9ea9f9ff4e; min_upper=d6e260afc48483aaa50547b09ac2071704931568; }
+1607624831us [INFO] irmin-pack cancelled freeze { id=7; waiting=53.707us; duration=1.236s }
+1607624831us [INFO] irmin-pack freeze started { id=8; max=dc67aa27d37ad5e392436a7a9a8a46ebcde73330; min_upper=6f26f9d7d46ebc525c06b52ca621aee01515056d; }
+1607624833us [INFO] irmin-pack cancelled freeze { id=8; waiting=99.128us; duration=2.548s }
+1607624833us [INFO] irmin-pack freeze started { id=9; max=73b5eeb4c5b913ff2768ce80f834b31f91b58a81; min_upper=de09f7fc557476f2695a66e07cb8bd05043290f2; }
+1607624835us [INFO] irmin-pack cancelled freeze { id=9; waiting=88.913us; duration=1.747s }
+1607624835us [INFO] irmin-pack freeze started { id=10; max=7f951531300211a33929bccba63fee0cd68c8231; min_upper=9f4c1195aa19aaf02b1ca9fb5c35d351b72d10fa; }
+1607624848us [INFO] irmin-pack freeze { id=10; waiting=133us; duration=13.595s }
+1607624848us  application 15000 commits completed in 29.02s.
[0.002s per commit, 517 commits per second]
dune exec -- ./bench/irmin-pack/layers.exe -n 1000 -v   25.21s user 4.53s system 93% cpu 31.919 total
```

As expected, only the last freeze completes and the waiting time for all freezes is minimal.

### CODE REVIEW

**Comment by zshipko** [comment]
Here are some memory benchmarks comparing this branch with master: https://github.com/mirage/irmin/issues/1184#issuecomment-742934650

**Review by icristescu — APPROVED** [review]
Looks good, thanks! 
I think it would be nice to have some tests on canceling freezes. 
Also a cancel should clear the objects that have already been copied; but maybe we handle that in a different PR.
And also I'm wondering how this will play with https://github.com/mirage/irmin/pull/1183 -in particular if we cancel a freeze after calling `Index.merge`.

File: src/irmin-pack/irmin_pack.ml [hunk 1]
```diff
@@ -30,7 +30,10 @@ exception RO_Not_Allowed = IO.Unix.RO_Not_Allowed
 
 exception Unsupported_version = Store.Unsupported_version
 
-type throttle = [ `Overcommit_memory | `Block_writes ] [@@deriving irmin]
+type index_throttle = [ `Overcommit_memory | `Block_writes ] [@@deriving irmin]
+
+type freeze_throttle = [ `Overcommit_memory | `Block_writes | `Cancel_existing ]
```
[the remark this answers is not in the document]
Comment (samoht) [reply src/irmin-pack/irmin_pack.ml:35]: unfortunately this suggestion breaks `ppx_irmin` as it cannot handle this kind of construct

File: src/irmin-pack/irmin_pack_layers.ml [hunk 2]
```diff
@@ -850,73 +867,82 @@ struct
...
-      pause () >>= fun () ->
-      may (fun f -> f `Before_Copy) hook >>= fun () ->
-      copy ~min ~max ~squash ~copy_in_upper ~min_upper t >>= fun () ->
+      copy ~cancel ~min ~max ~squash ~upper t >>= fun () ->
       X.Repo.flush_next_lower t;
       may (fun f -> f `Before_Copy_Newies) hook >>= fun () ->
       Copy.CopyToUpper.copy_newies_to_next_upper t offset >>= fun () ->
       may (fun f -> f `Before_Copy_Last_Newies) hook >>= fun () ->
       Lwt_mutex.with_lock t.add_lock (fun () ->
-          Log.debug (fun l -> l "freeze: enter blocking section");
```
[the remark this answers is not in the document]
Comment (samoht) [reply src/irmin-pack/irmin_pack_layers.ml:900]: I've added time information instead in #1192

File: src/irmin-pack/irmin_pack_layers.ml [hunk 3]
```diff
@@ -850,73 +867,82 @@ struct
...
           X.Repo.flip_upper t;
           may (fun f -> f `Before_Clear) hook >>= fun () ->
           X.Repo.clear_previous_upper t)
       >>= fun () ->
-      Log.debug (fun l -> l "freeze: exit blocking section");
       (* RO reads generation from pack file to detect a flip change, so it's
          ok to write the flip file outside the lock *)
-      X.Repo.write_flip t >>= fun () ->
-      Lock.close lock_file >>= fun () ->
+      X.Repo.write_flip t >>= fun () -> Lock.close lock_file
```
[the remark this answers is not in the document]
Comment (samoht) [reply src/irmin-pack/irmin_pack_layers.ml:918]: done in #1192

**Comment by samoht** [comment]
> Also a cancel should clear the objects that have already been copied; 

Do we want this? The good thing about keeping the old objects is that new freeze should be faster as the cancelled job already has done some work. I agree that in that case the GC won't be perfect but maybe that's fine. Need to think a bit more on this.

**Comment by samoht** [comment]
Rebased, but I haven't added the code to cleanup resources in case of cancellation. Not sure what to do exactly here, do we already have some code doing this?

File: src/irmin-pack/config.ml [hunk 4]
```diff
@@ -51,14 +51,33 @@ let throttle_converter : throttle Irmin.Private.Conf.converter =
   in
   (parse, print)
 
+type freeze_throttle = [ index_throttle | `Cancel_existing ]
```
Comment (icristescu) [inline src/irmin-pack/config.ml:54]: the configuration arguments for irmin_pack_layers are defined here https://github.com/mirage/irmin/blob/master/src/irmin-pack/irmin_pack_layers.ml#L55; but maybe we should move them all in `config.ml`?

**Comment by icristescu** [comment]
> Rebased, but I haven't added the code to cleanup resources in case of cancellation. Not sure what to do exactly here, do we already have some code doing this?

it should be similar to clearing the previous upper at the first freeze after a crash (using the [recovery flag](https://github.com/mirage/irmin/blob/master/src/irmin-pack/irmin_pack_layers.ml#L952)); we can clear the resources  either when the freeze is cancelled, or as for recovery, at the beginning of the next freeze.

**Comment by samoht** [comment]
Rebased and updated to:
- move the definition of config flags to `Irmin_pack.Config`
- clean-up the next layer when the operation is cancelled

**Pull request merged**